src/fromager/sdist.py

```diff
@@ -259,6 +278,150 @@ def handle_requirement(
+    req_type: RequirementType,
+    pre_built: bool,
+) -> tuple[str, Version] | None:
+    # we have already resolved top level reqs before bootstrapping
+    # so they should already be in the root node
+    if req_type == RequirementType.TOP_LEVEL:
+        for edge in ctx.dependency_graph.get_root_node().get_outgoing_edges(
+            req.name, RequirementType.TOP_LEVEL
+        ):
+            if edge.req == req:
+                return (
+                    edge.destination_node.download_url,
+                    edge.destination_node.version,
+                )
+        # this should never happen
+        return None
+
+    if not prev_graph:
+        return None
+
+    seen_version = set()
+
+    # first perform resolution using the top level reqs before looking at history
+    possible_versions_in_top_level = []
+    for top_level_edge in ctx.dependency_graph.get_root_node().get_outgoing_edges(
+        req.name, RequirementType.TOP_LEVEL
+    ):
+        possible_versions_in_top_level.append(
+            (
+                top_level_edge.destination_node.download_url,
+                top_level_edge.destination_node.version,
+            )
+        )
+        seen_version.add(str(top_level_edge.destination_node.version))
+
+    resolver_result = _resolver_for_graph(ctx, possible_versions_in_top_level, req)
+    if resolver_result:
+        return resolver_result
+
+    # only if there is nothing in top level reqs, resolve using history
+    possible_versions_from_graph = []
+    # check all nodes which have the same parent name irrespective of the parent's version
+    for parent_node in prev_graph.get_nodes_by_name(
+        parent_req.name if parent_req else None
+    ):
+        # if the edge matchesthe current req and type then it is a possible candidate
+        for edge in parent_node.get_outgoing_edges(req.name, req_type):
+            if (
+                edge.destination_node.pre_built == pre_built
+                and str(edge.destination_node.version) not in seen_version
+            ):
+                possible_versions_from_graph.append(
+                    (
+                        edge.destination_node.download_url,
+                        edge.destination_node.version,
+                    )
+                )
+                seen_version.add(str(edge.destination_node.version))
+
+    return _resolver_for_graph(ctx, possible_versions_from_graph, req)
+
+
+def _resolver_for_graph(
+    ctx: context.WorkContext, version_source: list[Requirement], req: Requirement
+):
+    if not version_source:
+        return None
+    try:
+        provider = resolver.GenericProvider(
+            version_source=lambda x, y, z: version_source, constraints=ctx.constraints
+        )
+        return resolver.resolve_from_provider(provider, req)
+    except Exception as err:
+        logger.debug(
+            f"{req.name}: could not resolve {req} from {version_source}: {err}"
+        )
+
+
+def _resolve_source(
```

Comment:
This name is very similar to the function in the sources module. It would be good to differentiate more. You could call it something like `_resolve_source_with_history` to emphasize that it looks at that graph for past resolutions.